### PR TITLE
fix(dashboard,widgets): correct silent aggregation errors in score histogram and pivot averages (#15208)

### LINKED ISSUE #15208: bug: silent wrong results in two score-analytics aggregations (histogram binning & pivot-table averages)
## Summary

While differential-testing Langfuse's client-side statistical code against reference libraries (`numpy`, `scikit-learn`), I found two **silent** correctness bugs — they return wrong numbers to users with no error, on the dashboards people use to read eval results:

1. **Numeric score histogram** (`createHistogramData`) counts scores in the **wrong bucket**, because it rounds each value to 2 decimals *before* binning. The rendered distribution disagrees with a standard histogram, and buckets are labeled with ranges that don't contain the values counted in them.
2. **Pivot-table subtotals/grand totals** (`calculateSubtotals`) compute averages as an **average of per-group averages** instead of a count-weighted average, so subtotal/total rows overstate or understate metrics like avg cost/latency whenever groups differ in size.

Both live in correctness-critical analytics paths. I've verified the neighbouring inter-rater stats (`calculateCohensKappa`, `calculateWeightedF1Score`, `calculateOverallAgreement`) against scikit-learn and they are **correct** (0 mismatches over 1,200 random cases) — these two are the outliers.

---

## Bug 1 — Numeric score histogram bins on rounded values

**File:** `web/src/features/dashboard/lib/score-analytics-utils.ts` → `createHistogramData` (rendered by the dashboard `NumericScoreHistogram`, produced in `dashboard-router.ts`).

The value is rounded to 2 decimals before its bin index is computed, while the bin edges are not:

```ts
const min = round(Math.min(...numericScoreValues));            // rounded
const binSize = range / bins || 1;
// ...
const shiftedValue = round(value) - min;                       // value rounded to 2dp
const binIndex = Math.min(Math.floor(shiftedValue / binSize), bins - 1);
```

### Reproduction (vs `numpy.histogram`)

A faithful port of the function was run against `numpy.histogram` (same bin count) over 1,200 random score datasets:

```
uniform 0..1 (n=50)          | value-in-wrong-bucket:185/200  numpy-mismatch:117/200
uniform 0..1 (n=200)         | value-in-wrong-bucket:200/200  numpy-mismatch:199/200
narrow 0.80..0.85 (n=40)     | value-in-wrong-bucket:200/200  numpy-mismatch:200/200
1..5 likert (n=60)           | value-in-wrong-bucket:  0/200  numpy-mismatch:  0/200
...
runs differing from numpy.histogram: 916 / 1200
```

Integer (Likert) scores are unaffected — which is why this has gone unnoticed — but any fractional scores (0–1 confidence, LLM-judge scores, cost, latency) are misbinned.

**Minimal deterministic case** (7 equal-width bins over `[0, 1]`, input `[0, 0.857, 0.86, 1]`):

| value  | bucket it's counted in (current) | bucket that contains it |
| ------ | -------------------------------- | ----------------------- |
| 0.857  | **`[0.86, 1]`** ❌               | `[0.71, 0.86]`          |

`round(0.857) = 0.86`, so `floor(0.86 / (1/7)) = 6` → the **top** bucket, even though `0.857 < 0.86`.

A second symptom of the same rounding: for narrow ranges the 2-decimal **labels collapse**, e.g. `createHistogramData([0.855, 0.857, 0.858, 0.861])` produces buckets labeled `["[0.85, 0.85]", "[0.85, 0.86]"]` — a zero-width bucket and duplicated edges.

### Fix

Bin on the raw value (`Math.floor((value - min) / binSize)`), derive edges from the raw min/max, and round only the **display labels** (with enough precision to keep edges distinct). After the fix the port matches `numpy.histogram` exactly (**0 / 1200** mismatches).

---

## Bug 2 — Pivot-table subtotals average the averages

**File:** `web/src/features/widgets/utils/pivot-table-utils.ts` → `applyAggregation` / `calculateSubtotals` (subtotal & grand-total rows in dashboard pivot-table widgets).

```ts
case "avg":
  return values.reduce((sum, val) => sum + val, 0) / values.length;   // mean of per-row means
```

Each pivot data row is already an aggregate (e.g. `avg_cost` over `count` items). Averaging those per-row averages **unweighted** only equals the true overall average when every group has the same size.

### Reproduction

The repo's own test data (and its docstring example) demonstrates it. `calculateGrandTotals` over:

```
{ count: 10, avg_cost: 0.5 }
{ count: 15, avg_cost: 0.6 }
{ count: 20, avg_cost: 0.2 }
{ count: 25, avg_cost: 0.4 }
{ count:  5, avg_cost: 0.8 }
```

| quantity                              | value      |
| ------------------------------------- | ---------- |
| **shown** (avg of averages) `Σavg/5`  | **0.5**    |
| **correct** (weighted) `Σ(avg·count)/Σcount = 32/75` | **0.4267** |

The dashboard silently overstates average cost by ~**17%** here. The existing test even encodes the wrong value:

```ts
avg_cost: 0.5, // (0.5 + 0.6 + 0.2 + 0.4 + 0.8) / 5
```

(`percentile` subtotals have the same problem and are already flagged as "a reasonable approximation" in the code.)

### Fix

Weight each row's value by its `count` when a count metric is present:
`Σ(avg_i · count_i) / Σ(count_i)`. This makes subtotal/total averages exact for the common case (a count column is present) and falls back to the previous unweighted behavior when no weights are available. The truly exact fix for percentiles is query-level (the database can recompute over raw rows), but count-weighting is a strict improvement.

---

## Methodology

Each function was ported 1:1 to JS and executed against reference implementations:
`numpy.histogram` (Bug 1), and `sklearn.metrics.{cohen_kappa_score, f1_score, accuracy_score}` for the neighbouring categorical stats (verified correct — not bugs). Repro scripts and a proposed patch with unit tests are available; happy to open a PR.

### PR DESCRIPTION
**Pull request by hassaanch23**
## What does this PR do?

Fixes two **silent** correctness bugs in client-side score-analytics aggregations — both return wrong numbers to users with no error. Found by differential-testing the functions against `numpy` / `scikit-learn`.

### 1. Numeric score histogram bins on rounded values

`web/src/features/dashboard/lib/score-analytics-utils.ts` → `createHistogramData` (rendered by the dashboard `NumericScoreHistogram`).

Each score was rounded to 2 decimals *before* its bin index was computed, while the bin edges were not — so values near a boundary were counted in the wrong bucket:

- A score of `0.857` was counted in the bucket **labeled `[0.86, 1]`** (`round(0.857) = 0.86`).
- Narrow ranges collapsed the 2-decimal labels into degenerate buckets like `"[0.85, 0.85]"`.
- A 1:1 port disagreed with `numpy.histogram` on **916 / 1200** random datasets.

**Fix:** bin on the raw value (`Math.floor((value - min) / binSize)`), derive edges from the raw min/max, and round only the display labels (with enough precision to keep edges distinct). The output now matches `numpy.histogram` exactly (0 / 1200). Integer/Likert scores are unchanged.

### 2. Pivot-table subtotals average the averages

`web/src/features/widgets/utils/pivot-table-utils.ts` → `calculateSubtotals` (subtotal & grand-total rows in pivot-table widgets).

Subtotal/grand-total averages were computed as an unweighted mean of the per-row averages, which only equals the true average when every group has the same size. On the repo's own test data the grand-total `avg_cost` was `0.5` where the count-weighted value is `32/75 = 0.4267` — a silent ~17% overstatement (the old test even encoded `// (0.5+0.6+0.2+0.4+0.8)/5`).

**Fix:** weight each row's value by a count column (exact for averages): the metric's own `count_<measure>`, else the row count (`count_count`), else the only count column. When several unrelated count columns leave no safe choice, or no weights are available, it falls back to the previous unweighted mean rather than letting metric order decide. (Percentile subtotals remain approximate — the exact fix there is query-level; count-weighting is a strict improvement.)

## Tests

- New `score-analytics-utils.clienttest.ts`: raw-magnitude binning, the `0.857` regression, no degenerate labels, count preservation, edge cases.
- Updated `pivot-table-utils.servertest.ts`: corrected the codified average-of-averages expectations to count-weighted values, plus a test that the weight follows each metric's own count column in either metric order, and falls back to the unweighted mean when the choice is ambiguous.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code.

## Checklist

- [x] PR title follows Conventional Commits.
- [x] Added/updated tests that prove the fixes (histogram + pivot).
- [x] Commented the non-obvious areas (why raw-magnitude binning; why count-weighting).
- [x] No documentation changes required.
- [ ] Lint and the full test suite are validated by CI on this PR.

Fixes #15208

<details open><summary><h3>Greptile Summary</h3></summary>

This PR fixes two silent client-side aggregation bugs in score analytics. Neither bug produced an error — both silently returned incorrect numbers to users.

- **Histogram binning (`score-analytics-utils.ts`)**: Values were rounded to 2 decimal places *before* bin index computation while edges were not, causing values near bin boundaries to land in the wrong bucket. The fix bins on raw values and applies rounding to display labels only, using a new `labelPrecision` helper to keep edges visually distinct for narrow ranges.
- **Pivot-table subtotals (`pivot-table-utils.ts`)**: Subtotal/grand-total averages were computed as an unweighted mean of per-row averages (correct only when all groups are equal-sized). The fix introduces `findWeightMetric` to locate the count column and `weightedMean` to weight each row's value by its cardinality, falling back to the previous behavior when no count column is present.
- Tests are updated and new: corrected test expectations now encode the count-weighted values, and a new histogram test file covers raw-magnitude binning, the specific 0.857 regression, count conservation, and edge cases.
</details>

<details><summary><h3>Confidence Score: 4/5</h3></summary>

Safe to merge; both fixes correct well-documented silent bugs with no new runtime risk introduced.

The histogram and weighted-average fixes are arithmetically correct and well-tested. The one remaining fragility is in `findWeightMetric`: when a pivot widget exposes more than one count-type column, the first match in the metrics array silently wins and could weight averages by the wrong count. This is unlikely in current data models but is not guarded against, so a second look at multi-count scenarios is worthwhile before this path is used more broadly.

`pivot-table-utils.ts` — specifically `findWeightMetric` and its interaction with `calculateSubtotals` when multiple count-style metrics are present in the same widget config.
</details>

<details><summary><h3>Flowchart</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    subgraph Histogram["createHistogramData (fixed)"]
        A[Raw values from DB] --> B[min/max from raw values]
        B --> C[binSize = range / bins]
        C --> D["labelPrecision(binSize) -> decimal count"]
        D --> E["Labels: round(edge, precision)"]
        A --> F["binIndex = floor((value - min) / binSize)"]
        F --> G[Clamp to bins-1]
        G --> H[Increment bin count]
    end

    subgraph Pivot["calculateSubtotals (fixed)"]
        I[Metrics list] --> J["findWeightMetric()"]
        J --> K{Count metric found?}
        K -->|avg or percentile| L["weightedMean(value x count)"]
        K -->|count/sum/min/max| M["applyAggregation() unchanged"]
        L --> N{Total weight > 0?}
        N -->|yes| O[Weighted result]
        N -->|no fallback| M
        M --> P[Round to 10dp]
        O --> P
    end
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    subgraph Histogram["createHistogramData (fixed)"]
        A[Raw values from DB] --> B[min/max from raw values]
        B --> C[binSize = range / bins]
        C --> D["labelPrecision(binSize) -> decimal count"]
        D --> E["Labels: round(edge, precision)"]
        A --> F["binIndex = floor((value - min) / binSize)"]
        F --> G[Clamp to bins-1]
        G --> H[Increment bin count]
    end

    subgraph Pivot["calculateSubtotals (fixed)"]
        I[Metrics list] --> J["findWeightMetric()"]
        J --> K{Count metric found?}
        K -->|avg or percentile| L["weightedMean(value x count)"]
        K -->|count/sum/min/max| M["applyAggregation() unchanged"]
        L --> N{Total weight > 0?}
        N -->|yes| O[Weighted result]
        N -->|no fallback| M
        M --> P[Round to 10dp]
        O --> P
    end
```

</a>
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 2 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 2
web/src/features/widgets/utils/pivot-table-utils.ts:528-533
**`findWeightMetric` may silently select the wrong count metric**

The second fallback — `m.toLowerCase().startsWith("count")` — is string-prefix matching only. If a widget exposes multiple count-style metrics (e.g. `count_traces` and `count_errors`, or `count` alongside `count_errors`), the first match in the `metrics` array wins. Weighting averages by `count_errors` instead of `count_traces` would produce silently incorrect subtotals with no indication of the mismatch. The function's JSDoc promises it returns "row cardinality", but the selection is ordering-dependent when more than one count column is present.

### Issue 2 of 2
web/src/features/dashboard/lib/score-analytics-utils.ts:51-53
**`labelPrecision` returns 2 for `binSize >= 1`, which can still produce misleading labels**

When all values are identical (range = 0), `binSize` is forced to `1` and `labelPrecision` returns `2`. The resulting label is `[round(value, 2), round(value+1, 2)]` — technically non-degenerate, but visually misleading: a single score of `0.5` shows as `[0.5, 1.5]`. Correctness is unaffected for non-degenerate cases, but the identical-values histogram label implies a 1-unit-wide range even though only one unique value exists — a UX artifact worth noting if the histogram is shown to end users.

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(dashboard,widgets): correct silent a..."](https://github.com/langfuse/langfuse/commit/65c60fae6f60ae01dc1d0ea88930c74d3f58fd2a) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=45313772)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

**Comment by hassaanch23** [comment]
Note for reviewers: #15211 fixes the same two bugs in the same four files — @Taranum01 and I arrived at this independently. I've left a comparison on #15208 (langfuse/langfuse#15208#issuecomment-5146143014) rather than here, since it concerns both PRs.

Short version: #15211's per-base `avg_<base>`↔`count_<base>` weight pairing is more correct than my global weight-metric lookup; my bin-size-derived label precision holds up better than a hardcoded 3 decimals on narrow ranges. Best outcome is one PR carrying both.

I'm fine with this one being closed in favour of #15211 — happy to move my two pieces across either way. Deferring to whichever a maintainer prefers.

**Comment by nimarb** [comment]
@claude review, be very allergic against tests which could be consolidated or left out in the final code

File: web/src/features/widgets/utils/pivot-table-utils.ts [hunk 1]
```diff
@@ -530,16 +564,41 @@ export function calculateSubtotals(
...
+      // fields coerce to 0 via extractMetricValues, as elsewhere in this file;
+      // a row with no count carries weight 0 and drops out of the weighted mean.
+      const pairs = data.map((row) => {
+        const extracted = extractMetricValues(row, [metric, weightMetric]);
+        return { value: extracted[metric], weight: extracted[weightMetric] };
+      });
+
+      const weighted = weightedMean(pairs);
+      result =
+        weighted ??
```
[the remark this answers is not in the document]
Comment (hassaanch23) [reply web/src/features/widgets/utils/pivot-table-utils.ts]: Fixed in cc90d098. The four assertions now expect the count-weighted values: `avg_latency` 100 and `p95_duration` 200 in the multi-metric grand total, `avg_latency` 93.6333 in the mixed string/number case, and 93.3333 in the two-dimension `production` subtotal. None of the old unweighted values (113.33, 233.33, 100.35) remain in the file.

File: web/src/__tests__/server/pivot-table-utils.servertest.ts [hunk 2]
```diff
@@ -353,7 +353,26 @@ describe("pivot-table-utils", () => {
...
+        { count: 5, avg_score: 60 },
+      ];
+
+      const result = calculateSubtotals(data, ["count", "avg_score"]);
+
+      expect(result).toEqual({
+        count: 15,
+        avg_score: 80,
       });
     });
```
[the remark this answers is not in the document]
Comment (hassaanch23) [reply web/src/__tests__/server/pivot-table-utils.servertest.ts:398]: Removed in 539c498c. "should calculate subtotals for numeric metrics" already proves the weighted path on its own, so the "weights averages by count instead of averaging the averages" case is gone. Its place is taken by a single test for the weight-selection change in the `findWeightMetric` thread, which covers a different path, so this block has the same number of tests as before.

File: web/src/features/widgets/utils/pivot-table-utils.ts [hunk 3]
```diff
@@ -517,6 +518,39 @@ function applyAggregation(values: number[], aggregationType: string): number {
...
+ * without weighting by each row's count yields an "average of averages", which
+ * only equals the true overall average when every group has the same size.
+ * Returns the count metric field, or null if the rows carry no count.
+ */
+function findWeightMetric(metrics: string[]): string | null {
+  return (
+    metrics.find((m) => detectAggregationType(m) === "count") ??
+    metrics.find((m) => m.toLowerCase().startsWith("count")) ??
+    null
+  );
```
[the remark this answers is not in the document]
Comment (hassaanch23) [reply web/src/features/widgets/utils/pivot-table-utils.ts:540]: Fixed in 539c498c. The weight is now chosen per metric: an average uses its own `count_<measure>` first, then the row count (`count_count`), then a count column only if it is the only one. If several unrelated count columns exist and none matches, it returns null and the subtotal falls back to the unweighted mean, so metric order can no longer change the result.

The new test checks that `[count_cost, count_latency, avg_cost, avg_latency]` gives identical subtotals in both orders, and that your `count_traces` / `count_observations` example gives 150, the unweighted mean. It fails on the previous code and passes now.

Because of the row-count and single-count fallbacks, the existing `count_requests` fixtures stay weighted, so the four corrected assertions are unaffected.

[the remark this answers is not in the document]
Comment (hassaanch23) [reply web/src/features/widgets/utils/pivot-table-utils.ts]: Addressed in 539c498c. Instead of taking the first count-like column for the whole table, each average is weighted by its own `count_<measure>`, then by the row count, then by a count column only if it is the only one. If the choice is still ambiguous, it falls back to the unweighted mean.

File: web/src/features/dashboard/lib/score-analytics-utils.ts [hunk 4]
```diff
@@ -44,6 +44,15 @@ function computeBinSize(
     : Math.min(Math.max(minBins, Math.floor(Math.sqrt(valueCount))), maxBins);
 }
 
+// Decimals needed so adjacent bin edges stay distinct in labels. Values are
+// binned on their raw magnitude (see below); rounding is display-only. Without
+// this, small ranges collapse edges to the same 2dp value and produce
+// degenerate labels like "[0.85, 0.85]".
+function labelPrecision(binSize: number): number {
+  if (!Number.isFinite(binSize) || binSize <= 0 || binSize >= 1) return 2;
+  return Math.min(10, Math.max(2, -Math.floor(Math.log10(binSize)) + 1));
```
[the remark this answers is not in the document]
Comment (hassaanch23) [reply web/src/features/dashboard/lib/score-analytics-utils.ts:49]: Agreed that it reads oddly, and that it predates this PR: `main` renders the same `[0.5, 1.5]` when every value is identical. I've kept it out so this PR stays limited to the two aggregation fixes. Happy to send a separate change if you want it, for example a single-value label when the range is zero.

**Comment by hassaanch23** [comment]
@nimarb I've addressed the 20 Aug review in 539c498c and merged current `main`:

- The four stale pivot assertions were already corrected in cc90d098.
- I removed the duplicate weighting test.
- `findWeightMetric` now weights each average by its own `count_<measure>`, then by the row count, then by a count column only if it is the only one. When that's still ambiguous it uses the unweighted mean, so metric order can no longer change a subtotal. There's a new test covering both orders and the ambiguous case.

The degenerate-histogram label predates this PR, so I've left it out of scope; details are in that thread. CI on this fork is waiting for workflow approval.

**Comment by hassaanch23** [comment]
Gentle ping on this one — it's been open since 18 July, and the test and lint workflows haven't run yet because fork PRs need a maintainer to approve them. So far only the CLA check has run (signed).

Quick context for whoever picks it up:

- Two silent wrong-result bugs in score analytics (#15208): histogram values were rounded *before* binning, so scores near a bin edge landed in the wrong bucket; and pivot-table subtotals averaged the per-group averages without weighting by count.
- #15211 fixes the same issue. I compared the two in #15208 — each is stronger in a different place. I'm happy to fold the better parts of #15211 into this PR, or to close this one in its favour. I'd just rather you didn't end up reviewing the same fix twice.

A CI approval would at least let this be judged on a green run.
